Update OpenTracingTracer to support api deprecations/removals in opentracing 0.33.0

mahinshaw opened the pull request:
The `OpenTracingTracer` was updated to use Spans instead of scopes as the entry point. This fixes the deprecations and removals to the api. I have started work on a ScopeManager that uses `Context` as defined by Vertx as opposed to  `ThreadLocal` state. I need to add some tests for that before I contribute it. Would it make sense to add that in another PR, or add it to this PR?

mahinshaw commented:
I've added a ScopeManager that leverages Vertx's Context. The calls made to the VertxTracer have cases where the inbound Context is not attached to a Thread (VertxThread more specifically). This creates a situation where `Vertx.currentContext()` will return null, as it looks for the context that is attached to the active thread. That lead to the static `withContext` method, which allows for a closable state to attain and leverage the passed Context.  I also added the VertxContextScopeManager to the default Tracer that is created when the TracingOptions are configured.

vietj commented:
can we avoid using mockito ?

vietj commented:
@pavolloffay can you have a look at this PR ?

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/OpenTracingTracer.java:
```diff
@@ -48,88 +43,94 @@ public OpenTracingTracer(boolean closeTracer, Tracer tracer) {
+        .ignoreActiveSpan()
+        .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
+        .withTag(Tags.COMPONENT.getKey(), "vertx").start();
+      tracer.activateSpan(span);
```
The scope should not be activated here. There is no method which returns the scope anymore, so if you do not pass it around there will be a leak

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/OpenTracingTracer.java:
```diff
@@ -48,88 +43,94 @@ public OpenTracingTracer(boolean closeTracer, Tracer tracer) {
-        });
+  public <R> Span sendRequest(Context context, R request, String operation,
+    BiConsumer<String, String> headers, TagExtractor<R> tagExtractor) {
+    try (CapturedContext ignored = VertxContextScopeManager.withContext(context)) {
```
See my other comment, you should perhaps store the span in vertx context and get it from there as opposed to `activeSpan()`

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/VertxContextScopeManager.java:
```diff
@@ -0,0 +1,114 @@
+import io.vertx.core.Vertx;
+
+/** VertxContextScopeManager is based on the {@link io.opentracing.util.ThreadLocalScopeManager}. */
+public class VertxContextScopeManager implements ScopeManager {
```
Can we get around without this? I ideal would be to use only spans and where needed active them in the blocking execution to make them available in user-defined handlers.

mahinshaw replied:
To be clear, you're advocating for removing the idea of a Context based scope manager? In implementing this, I was under the assumption that, given the potential for multiple contexts  on a given thread, it would be better to attach scopes to a context. Am I missing something?

mahinshaw commented on vertx-opentracing/pom.xml:
```diff
@@ -36,18 +36,14 @@
       <groupId>io.opentracing</groupId>
       <artifactId>opentracing-mock</artifactId>
-      <version>0.31.0</version>
+      <version>0.33.0</version>
```
I'll fix that.

mahinshaw commented:
@pavolloffay Thanks for the feedback. I am trying to understand the comments about the `activeSpan`. I was assuming that the scope manager would allow for the user to create an manage scopes that leave a context/thread/etc. If all spans are attached to the Context directly, then would the user not need to access the `Context` throughout their code? If that is the case, is there a good way to provide them access to the active context without leveraging a ScopeManager? It seems that java-grpc does something similar with  [the current span context](https://github.com/opentracing-contrib/java-grpc#current-span-context) call. I understand that `Vertx.currentContext` can be called, but that feels a little heavy.

mahinshaw commented:
Ok, after reading through the linked ticket, I think I have a better understanding of what you mean by avoiding the Context based scope manager. I will update the PR with the requested changes. Also, I will add a static handler to the OpenTracingTracer to access the context for an active span like the grpc-java library does.

mahinshaw commented:
@pavolloffay I think I addressed your points. I removed the ScopeManager implementation, and no longer set Spans as active in the OpenTracingTracer. Also, I added `OpenTracingContext` with some static helpers for accessing state on the current context.

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/OpenTracingContext.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class OpenTracingContext {
+
+  static final String ACTIVE_SPAN = "vertx.tracing.opentracing.span";
```
why not private?

mahinshaw replied:
I needed access to the string in the OpenTracingTracer. For why, see my comment below.

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/OpenTracingContext.java:
```diff
@@ -0,0 +1,48 @@
+ * OpenTracingContext adds helpers for associating and disassociating spans with the current {@link
+ * Context}
+ */
+public class OpenTracingContext {
```
final?

mahinshaw replied:
I'll fix that.

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/OpenTracingTracer.java:
```diff
@@ -48,37 +46,35 @@ public OpenTracingTracer(boolean closeTracer, Tracer tracer) {
+      .withTag(Tags.COMPONENT.getKey(), "vertx")
+      .start();
+    reportTags(span, request, tagExtractor);
+    context.putLocal(ACTIVE_SPAN, span);
```
Use the `OpenTracingContext` class

mahinshaw replied:
I didn't use the util class here, because `Vertx.currentContext()` uses the Context from the active `VertxThread`. In the case of this call, the Context that is being passed to the method is not attached to the current `VertxThread`. I ran into this issue when I was working on the ScopeManager implementation. It lead me to adding the `withContext` method, which was awkward at best. So for the OpenTracingTracer class, I used the context directly, and the Util class will be merely helper methods for "user" code.

pavolloffay replied:
I see, you could have a helper method in the context utils class which accepts the context. I am fine with this if we expect folks will not need the same method in the utils.

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/OpenTracingContext.java:
```diff
@@ -0,0 +1,48 @@
+ * OpenTracingContext adds helpers for associating and disassociating spans with the current {@link
+ * Context}
+ */
+public class OpenTracingContext {
```
I would rename it to util as it is not context implementation. E.g. `OpenTracingContextUtisl`

mahinshaw replied:
Will fix

pavolloffay commented on vertx-opentracing/src/main/java/io/vertx/tracing/opentracing/OpenTracingContext.java:
```diff
@@ -0,0 +1,48 @@
+   *
+   * @return a {@link Span} or null
+   */
+  public static Span activeSpan() {
```
I don't like the `active`, `activate` names here as it can be confused with ScopeManagers. Consider using `set`/`setSpan`.. `get`/...

mahinshaw replied:
Ok. I'll change those.

mahinshaw commented:
@pavolloffay I believe I've addressed your comments. Let me know if there is anything else. Thanks for the review!

vietj commented:
thanks for the review @pavolloffay and your contribution @mahinshaw
